### PR TITLE
[Feat] Added forget password feature

### LINKED ISSUE #2: Add Forgot password functionality in login page
allow user to reset his password if he/she forgets his password

### PR DESCRIPTION
**Pull request by SurajPrakash24**
## 📋 Description
Added the forgot-password feature to the login page making the user authentication more efficient.

## 🔨 Changes Made

Fixes #2 [Feat] - Forgot password functionality in login page

-  Added API endpoints for forgot-pass and reset-pass
-  OTP verification via Email to reset the password
-  Created two components and two pages for forgot-Pass and reset-Pass
-  Real-time user feedback with react toast messages
-  updated the User model

## 🏷️ Types of Changes

What type of changes does your code introduce?
- [ ] New feature (non-breaking change which adds functionality) ✨

## 📸 Screenshots (if applicable)
login
forgot-pass
email2
gmail
reset-pass

---
@Sibam-Paul  Please accept it under gssoc-ext level 3 and hacktoberfest as it has taken reasonable amount of time and efforts.
Thank You !!!

### CODE REVIEW

**Comment by SurajPrakash24** [comment]
@Sibam-Paul Please review the PR and merge it under gssoc-ext level3 and hacktoberfest.

**Comment by Sibam-Paul** [comment]
@SurajPrakash24 there is conflict can u please update ur branch

File: backend/models/userModel.js [hunk 1]
```diff
@@ -17,6 +17,11 @@ const userSchema = mongoose.Schema(
       required: true,
       default: false,
     },
+    resetOtp: {
```
Comment (kartikayasijaa) [inline backend/models/userModel.js:20]: Why you added this otp field in User Table?

File: backend/controllers/userControllers.js [hunk 2]
```diff
@@ -71,8 +75,81 @@ const authUser = asyncHandler(async (req, res) => {
...
+  const { email } = req.body;
+
+  try {
+    const user = await User.findOne({ email });
+    if (!user) {
+      res.status(400);
+      throw new Error("User with this email does not exist");
+    }
+    // Generate token
+    const generateOtp = () => {
```
Comment (kartikayasijaa) [inline backend/controllers/userControllers.js:92]: move this function in utils/helpers

File: backend/controllers/userControllers.js [hunk 3]
```diff
@@ -71,8 +75,81 @@ const authUser = asyncHandler(async (req, res) => {
...
+      service: 'gmail',
+      auth: {
+        user: process.env.EMAIL_USERNAME,
+        pass: process.env.EMAIL_PASSWORD,
+      },
+    });
+
+    const otp = resetToken;
+    const mailOptions = {
+      from: process.env.EMAIL_USERNAME,
```
Comment (kartikayasijaa) [inline backend/controllers/userControllers.js:112]: handle the case if there's no email_username in env

File: backend/controllers/userControllers.js [hunk 4]
```diff
@@ -71,8 +75,81 @@ const authUser = asyncHandler(async (req, res) => {
...
+    throw new Error("Server Error");
+  }
+});
+
+const resetPass = asyncHandler(async (req, res) => {
+  const { otp, password } = req.body;
+
+  try {
+    const user = await User.findOne({
+      resetOtp: otp,
```
Comment (kartikayasijaa) [inline backend/controllers/userControllers.js:133]: Make a new model for this OTP store.

File: backend/controllers/userControllers.js [hunk 5]
```diff
@@ -71,8 +75,81 @@ const authUser = asyncHandler(async (req, res) => {
...
+      resetOtpExpires: { $gt: Date.now() }
+    });
+
+    if (!user) {
+      return res.status(400).send('OTP is invalid or has expired.');
+    }
+
+    // Update password
+    user.password = password;
```
Comment (kartikayasijaa) [inline backend/controllers/userControllers.js:142]: hash the password, see how we did it in signup

File: frontend/src/components/Authentication/ForgotPassword.js [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+import { axiosReq } from "../../config/axios";
+import { useNavigate } from "react-router-dom";
+
+const ForgotPassword = () => {
+  const [email, setEmail] = useState();
+  const [loading, setLoading] = useState(false);
+  const toast = useToast();
+  const navigate = useNavigate();
+
+  const submitHandler = async () => {
```
Comment (kartikayasijaa) [inline frontend/src/components/Authentication/ForgotPassword.js:16]: make a custom hook, and move this logic to that.

File: frontend/src/Pages/ResetPassPage.js [hunk 7]
```diff
@@ -0,0 +1,28 @@
+import { Container, Box, Text } from "@chakra-ui/react";
+import ResetPassword from "../components/Authentication/ResetPassword";
+
+function ResetPasswordPage() {
+  return (
+    <Container maxW="xl" centerContent>
```
Comment (kartikayasijaa) [inline frontend/src/Pages/ResetPassPage.js:6]: Why forget and reset password is different?

File: backend/controllers/userControllers.js (hunk 2 above)
Comment (SurajPrakash24) [reply backend/controllers/userControllers.js:92]: Ok

File: backend/controllers/userControllers.js (hunk 3 above)
Comment (SurajPrakash24) [reply backend/controllers/userControllers.js:112]: Ok

File: backend/controllers/userControllers.js (hunk 4 above)
Comment (SurajPrakash24) [reply backend/controllers/userControllers.js:133]: Ok

File: backend/models/userModel.js (hunk 1 above)
Comment (SurajPrakash24) [reply backend/models/userModel.js:20]: I will make another table to store otp.

File: frontend/src/Pages/ResetPassPage.js [hunk 8]
```diff
@@ -0,0 +1,28 @@
+import { Container, Box, Text } from "@chakra-ui/react";
+import ResetPassword from "../components/Authentication/ResetPassword";
```
[the remark this answers is not in the document]
Comment (SurajPrakash24) [reply frontend/src/Pages/ResetPassPage.js:2]: Both the components have different fields in forms. resetPass has Otp, new password, confirm password fields while forget-pass has only email input field.

File: frontend/src/Pages/ResetPassPage.js (hunk 7 above)
Comment (SurajPrakash24) [reply frontend/src/Pages/ResetPassPage.js:6]: forget-password is to take email as input and then after verifying the email exists in db, it sends the otp to that email.
reset-password is to verify otp and then take new password as input from user. And after sending the email, I am navigating to the rest-pass page from forget-pass page.

**Comment by SurajPrakash24** [comment]
@kartikayasijaa I have made all the required changes and committed to this PR.

1. Created separate table to store OTP related fields
2. Created separate files to generateOTP & emailService and added in utils/helpers directory
3. handle the case if there is no  email_username or  email_password in env file
4. For hashing the password, there is already a pre save middleware to hash the password before saving it.
5. forgot-password and reset-password have different fields so there is different component for each

Plz review and merge it under gssoc-ext level 3 and hacktoberfest.

**Comment by Sibam-Paul** [comment]
@SurajPrakash24 Please update ur branch and make the changes

**Comment by SurajPrakash24** [comment]
> @SurajPrakash24 Please update ur branch and make the changes

It is showing as there is no conflicts.

**Comment by SurajPrakash24** [comment]
@Sibam-Paul plz merge the PR under gssoc-ext level3 and hacktoberfest-accepted.

**Comment by SurajPrakash24** [comment]
@Sibam-Paul plz review and merge the PR.

File: backend/controllers/userControllers.js [hunk 9]
```diff
@@ -58,6 +64,15 @@ const authUser = asyncHandler(async (req, res) => {
   const { email, password } = req.body;
 
   const user = await User.findOne({ email });
+<<<<<<< HEAD
```
Comment (Sibam-Paul) [inline backend/controllers/userControllers.js:67]: what is this please fix this 
image

File: backend/controllers/userControllers.js [hunk 10]
```diff
@@ -58,6 +64,15 @@ const authUser = asyncHandler(async (req, res) => {
...
+
+=======
+  
+>>>>>>> be56316e4935f90c182df5aefe35f49a868ca9e0
```
Comment (Sibam-Paul) [inline backend/controllers/userControllers.js:71]: remove the dummy lines and text

File: frontend/src/Pages/ResetPassPage.js (hunk 7 above)
Comment (Sibam-Paul) [reply frontend/src/Pages/ResetPassPage.js:6]: ok

**Comment by Sibam-Paul** [comment]
image

**Comment by Sibam-Paul** [comment]
@SurajPrakash24 please tell getting error in testing

File: backend/.env.sample [hunk 11]
```diff
@@ -1,4 +1,11 @@
...
+# JWT_SECRET=secret 
+# MONGO_URL=mongodb://127.0.0.1:27017/talk-trove 
+# PORT=5000
+# NODE_ENV=dev
+
+JWT_SECRET="Anything"
+MONGO_URL="mongodb://127.0.0.1:27017/talk-trove" #default
+PORT=5000 #default
+NODE_ENV=""
+EMAIL_USERNAME="" #sender email
```
Comment (Sibam-Paul) [inline backend/.env.sample:10]: what is this for Email username and password

**Review by Sibam-Paul — COMMENTED** [review]
Please make the review and make a working video

**Comment by SurajPrakash24** [comment]
@Sibam-Paul have you given the email credentials in .env file.
Also by default Gmail blocks using the password for less secure application(here nodemailer). So you have to create application specific password.

**Comment by SurajPrakash24** [comment]
Steps for creating application specific password:

Enable 2-Step Verification and generate Application password, then you can use the generated password to send emails using nodemailer.To do so you need to do the following:

- Go to your Google account at https://myaccount.google.com/
- Go to Security
- Choose 2-Step Verification - here you have to verify yourself, in my case it was with phone number and a confirmation code send as text message. After that you will be able to enabled 2-Step Verification
- Visit https://myaccount.google.com/apppasswords to create your app.
- Put a name e.g. nodemailer to your app and create it.
- A modal dialog will appear with the password. Get that password and use it in your code.

File: backend/.env.sample (hunk 11 above)
Comment (SurajPrakash24) [reply backend/.env.sample:10]: These fields are sender email credentials. For sending the email containing OTP to reset password, this email address is used as sender.
For EMAIL_USERNAME just give email address And for EMAIL_PASSWORD, you have to create application specific password from your Google account. I have given the steps below.

**Comment by Sibam-Paul** [comment]
@SurajPrakash24 please join the discord see discussions tab and message me there

**Pull request merged**